### ISSUE TITLE
Pause between downloads?

### ISSUE THREAD

**Issue by Retrograde-i486** [dewittpe/R-install-dependencies#2]
Hey I wanted to say I love this script, it helps me out a ton pulling down R packages for my team. 
I had a question about downloads, is there a way I can tell it to pause a few seconds between download requests? I have to go through a proxy and sometimes it'll just hang on a random package which causes the package download to timeout and then everything and I have to rerun the script. I'm wondering if there is a rate limit to cran.rstudio.com and that's why it happens.

I'm not much of a coder but I tried adding a line that says Sys.sleep(5) in a few places in the script but can't manage to get it to pause between each download.

**Comment by dewittpe** [comment]
Hi @Retrograde-i486, I'm happy to know you found this script helpful.   

All the downloads are currently happening in one call to [`download.packages`](https://github.com/dewittpe/R-install-dependencies/blob/826fa75781bc8a4c0c78bf1558030ade97208e97/build-dep-list.R#L102-L106).  To inject a pause between downloads this one call would need to replaced with a loop, something like this:

```r
dwnld_pkgs <- NULL
for(pkg in pkgs_to_download) {
  message("pausing five seconds before next download....")
  Sys.sleep(5)
  dwnld_pkgs <-
    rbind(dwnld_pkgs,
          download.packages(pkgs = pkg,
                            destdir = "pkg-source-files",
                            repos = c(CRAN, BIOC),
                            type = "source")
    )
}
```
 
As for the download hanging on random packages.  I've run into this before too.  The cases I've run into have been the result of a network security screening finding something in the package to prevent a download, or two much data in one download.  In these cases I've been able to get the downloads I need by disconnecting from VPNs or switching to a different network.  In your case, if the proxy is the issue, and switching networks isn't an option, I'm not sure what to suggest if adding a pause between downloads doesn't help.   

You could also try switching CRAN mirrors by changing the urls [defined in the script](https://github.com/dewittpe/R-install-dependencies/blob/826fa75781bc8a4c0c78bf1558030ade97208e97/build-dep-list.R#L63-L65).

This does show that the script could be improved significantly to account for partial downloads.

**Comment by Retrograde-i486** [comment]
Hey Peter,

Thanks so much for the code snippet, I added it into the script in what I believe is the correct place but getting different than expected behavior. 

I run the script and there's a brief pause where I think traffic is connecting to the proxy. Then I get the message I put in "Pausing 5 seconds before next download..."
Then it proceeds to start downloading the packages in my list without pausing until I get to one where it hangs as per usual. So I may be putting your code in the wrong place but just thought I'd mention it before I go back to poking at it.

Oh yeah, regarding the vpn/switching networks, definitely not an option haha. I have some other proxy servers we use that i've been trying. Had my network person in charge of the proxies check the traffic and they can see it going through correctly.

**Comment by dewittpe** [comment]
@Retrograde-i486 I have put a version of [the script](https://github.com/dewittpe/R-install-dependencies/blob/574a95d4460553020e2b8bdc22389c5e6fbb8334/build-dep-list.R) with the paused download on a new branch in the repo.  check it out and see if it works the way you were hoping for.

**Comment by Retrograde-i486** [comment]
Thank you so much for the update script, it's definitely pausing between each download now. I've played around with timings (3 and 5 seconds) but still running into that "SSL connect error" even with the pause but I'll work with our proxy and firewall people and see if there's something there.

I assume resuming downloads is pretty complicated but I've definitely noticed that when you re-run the script it wipes out anything downloaded into pkg-source-files

**Comment by dewittpe** [comment]
@Retrograde-i486, I've updated the script to do three things:

1. Check for the version of a downloaded package and only download the source file if there the package has not been download or if there is a newer version available.
2. Allow for pausing between downloads
3. Improved Makefile

If you can test it out and let me know if it works for you I would appreciate it.

**Comment by Retrograde-i486** [comment]
Sorry for the delay, just got back from vacation. Tested the new script and it worked perfectly! Gave it a list of of 318 R packages and it pulled down the majority with no problem. The few that it did timeout on I just re-ran the script and saw the messages about it skipping the ones it already had and moved onto the failed ones. Only had to do that once and I got everything. Thank you so much, script is even more awesome now, if I could sponsor this project I would!
